src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,91 @@
+                            // throw an ExtensionError.
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
+                            let eventType = await new Promise((resolve, reject) => {
+                                tcp_socket.onopen = ((event) => {
+                                    resolve(event.type);
+                                });
+
+                                tcp_socket.onerror = ((event) => {
+                                    throw new ExtensionError(event.name);
+                                });
+                            });
+                            if (eventType != "open" || tcp_socket.readyState != "open") {
+                                throw new ExtensionError("TCP socket didn't properly open");
+                            }
+
+                            // After we know that the socket is open, send the bytes
+                            tcp_socket.send(buf.buffer, buf.byteOffset, buf.byteLength);
+                            let responseBytes = await new Promise((resolve, reject) => {
+                                let data = new Uint8Array();
+                                let expectedLength;
+
+                                tcp_socket.ondata = ((event) => {
+                                    if (data.length == 0) {
+                                        expectedLength = new DataView(event.data).getUint16(0);
+                                        let receivedData = new Uint8Array(event.data);
+                                        data = receivedData;
+                                    } else {
+                                        data = concatUint8Arrays(data, receivedData);
+                                    }
+
+                                    // Check if we have got all the expected data, or if we've got too much data
+                                    if (data.length == expectedLength + 2) {
+                                        resolve(data);
+                                    } else if (data.length > expectedLength + 2) {
+                                        throw new ExtensionError("Got too many bytes from TCP");
+                                    }
+                                });
+
+                                tcp_socket.onopen = ((event) => {
+                                    throw new ExtensionError("Got an additional open event");
+                                });
+
+                                tcp_socket.onerror = ((event) => {
+                                    throw new ExtensionError(event.name);
+                                });
+                            });
+                            return responseBytes;
+                        } catch(e) {
+                            throw new ExtensionError(e.message);
+                        } finally {
+                            tcp_socket.close();
```

Comment:
To ensure that the socket is closed as soon as the extension is unloaded, do this:

After creating a socket instance:

```
let closeHandler = {
  close() {
    try {
      tcp_socket.close();
    } catch (e) {
      Cu.reportError(e);
    }
  },
};
context.callOnClose(closeHandler);
```

In finally:

```
context.forgetOnClose(closeHandler);
closeHandler.close();
```

Reply:
Addressed in 62e4fc5